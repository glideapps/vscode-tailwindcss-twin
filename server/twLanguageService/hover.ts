import { serializeError } from "serialize-error"
import * as lsp from "vscode-languageserver"
import { TextDocument } from "vscode-languageserver-textdocument"
import { canMatch, PatternKind } from "~/common/ast"
import { hoverElement } from "~/common/findElement"
import parseThemeValue from "~/common/parseThemeValue"
import * as tw from "~/common/token"
import { Tailwind } from "~/tailwind"
import type { ServiceOptions } from "~/twLanguageService"
import { arbitraryValueDocs } from "./arbitraryValue"
import { cssDataManager, getEntryDescription } from "./cssData"
import { renderClassname, renderClassnameJIT, renderVariant } from "./markdown"
import { getDescription, getReferenceLinks } from "./referenceLink"

export default async function hover(
	document: TextDocument,
	position: lsp.Position,
	state: Tailwind,
	options: ServiceOptions,
): Promise<lsp.Hover | undefined> {
	try {
		const result = canMatch(document, position, true, options.jsxPropImportChecking)
		if (!result) {
			return undefined
		}
		const { token, kind } = result
		if (kind === PatternKind.TwinTheme) {
			const range = {
				start: document.positionAt(token.start),
				end: document.positionAt(token.end),
			}
			return resolveThemeValue({ kind, range, token, state, options })
		} else {
			const selection = hoverElement({
				input: token.text,
				position: document.offsetAt(position) - token.start,
				separator: state.separator,
			})
			if (!selection.token) {
				return undefined
			}

			const { start, end, text } = selection.token.token
			const range = lsp.Range.create(
				document.positionAt(token.start + start),
				document.positionAt(token.start + end),
			)

			if (selection.token.kind === tw.TokenKind.CssProperty) {
				if (!arbitraryValueDocs[selection.token.key.text]) {
					const key = selection.token.key.toKebab()
					const value = selection.token.value.text
					const important = selection.important

					const values = [
						"```scss",
						"& {",
						`\t${key}: ${tw.formatCssValue(tw.removeComments(value))}${important ? " !important" : ""};`,
						"}",
						"```\n",
					]

					const entry = cssDataManager.getProperty(key)
					if (entry) {
						values.unshift(getEntryDescription(entry, true).value + "\n\n---\n")
					}

					return {
						range,
						contents: {
							kind: lsp.MarkupKind.Markdown,
							value: values.join("\n"),
						},
					}
				}
			}

			if (
				selection.token.kind === tw.TokenKind.ClassName &&
				!state.twin.isSuggestedClassName(selection.variants.texts, text)
			) {
				return undefined
			}

			if (selection.token.kind === tw.TokenKind.CssProperty) {
				if (kind !== PatternKind.Twin || !arbitraryValueDocs[selection.token.key.text]) return undefined
			}

			let markdown: lsp.MarkupContent | undefined

			if (selection.token.kind === tw.TokenKind.CssProperty) {
				const className = text.replace(/\s/g, "")
				const twin = await state.jit(className)
				const content = renderClassnameJIT({
					key: className,
					twin,
					important: selection.important,
					options,
				})
				if (content) {
					markdown = { kind: lsp.MarkupKind.Markdown, value: content }
				}
			} else {
				markdown = getHoverTwinMarkdown({
					kind,
					selection,
					state,
					options,
				})
			}

			if (!markdown) {
				return undefined
			}

			let keyword = ""
			if (selection.token.kind === tw.TokenKind.CssProperty) {
				keyword = selection.token.key.text
			} else {
				text.replace(new RegExp(`^${state.config.prefix}`), "").replace(state.separator, "")
			}

			let title = ""
			if (options.references) {
				const type = getDescription(keyword)
				if (type) {
					title = type + "\n"
				}

				if (type === "") {
					title = "twin.macro" + "\n"
				}

				const refs = getReferenceLinks(keyword)
				if (refs.length > 0) {
					title += "\n" + refs.map(ref => `[Reference](${ref.url}) `).join("\n") + "\n"
				}
			}

			if (title) {
				markdown.value = `${title}\n---\n\n` + markdown.value
			}

			if (!title && options.references) {
				markdown.value = "custom\n" + "\n---\n\n" + markdown.value
			}

			return {
				range,
				contents: markdown,
			}
		}
	} catch (err) {
		console.log(serializeError(err))
	}
	return undefined
}

function getHoverTwinMarkdown({
	kind,
	selection,
	state,
	options,
}: {
	kind: PatternKind
	selection: ReturnType<typeof hoverElement>
	state: Tailwind
	options: ServiceOptions
}): lsp.MarkupContent | undefined {
	const { important } = selection

	if (selection.token == undefined) {
		return undefined
	}

	const value = selection.token.token.text

	if (selection.token.kind === tw.TokenKind.Variant) {
		const content = renderVariant({ state, key: value })
		if (content) {
			return { kind: lsp.MarkupKind.Markdown, value: content }
		}
		return undefined
	}

	if (selection.token.kind !== tw.TokenKind.ClassName) {
		return undefined
	}

	const content = renderClassname({ state, key: value, important, options })
	if (content) {
		return { kind: lsp.MarkupKind.Markdown, value: content }
	}
	return undefined
}

function resolveThemeValue({
	range,
	token,
	state,
}: {
	kind: PatternKind
	range: lsp.Range
	token: tw.Token
	state: Tailwind
	options: ServiceOptions
}): lsp.Hover | undefined {
	const result = parseThemeValue(token.text)
	if (result.errors.length > 0) {
		return undefined
	}

	const value = state.getTheme(result.keys())

	const markdown: lsp.MarkupContent = {
		kind: lsp.MarkupKind.Markdown,
		value: "",
	}

	if (typeof value === "string") {
		markdown.value = `\`\`\`txt\n${value}\n\`\`\``
	} else if (value instanceof Array) {
		markdown.value = `\`\`\`txt\n${value.join(", ")}\n\`\`\``
	} else if (value) {
		markdown.value = `\`\`\`js\n${value.toString?.() ?? typeof value}\n\`\`\``
	}

	return {
		range,
		contents: markdown,
	}
}
