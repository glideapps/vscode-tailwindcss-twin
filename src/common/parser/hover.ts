import * as nodes from "./nodes"
import { parse } from "./parse_regexp"
import { getVariant } from "./util"

interface HoverResult {
	target:
		| nodes.SimpleVariant
		| nodes.ArbitraryVariant
		| nodes.Classname
		| nodes.ArbitraryClassname
		| nodes.CssDeclaration
	value: string
	variants: string[]
	important: boolean
}

export function hover({
	position,
	text,
	start = 0,
	end = text.length,
}: {
	position: number
	text: string
	start?: number
	end?: number
}): HoverResult | undefined {
	interface Context {
		important: boolean
		variants: string[]
	}

	const inRange = (node: nodes.Node) => position >= node.range[0] && position < node.range[1]

	const travel = (node: nodes.Node, ctx: Context): HoverResult | undefined => {
		const walk = (node: nodes.Node): HoverResult | undefined => {
			if (nodes.NodeType.VariantSpan === node.type) {
				const variants = ctx.variants.slice()
				if (inRange(node.variant)) {
					return {
						target: node.variant,
						value: getVariant(node.variant).value,
						important: false,
						variants,
					}
				}
				if (!node.child) return undefined
				variants.push(getVariant(node.variant).value)
				return travel(node.child, { ...ctx, variants })
			}

			if (nodes.NodeType.Group === node.type) {
				return travel(
					{ type: nodes.NodeType.Program, expressions: node.expressions, range: node.range },
					{ ...ctx, important: ctx.important || node.important },
				)
			}

			if (nodes.NodeType.ClassName === node.type) {
				return {
					target: node,
					value: text.slice(node.range[0], node.range[1]),
					important: ctx.important || node.important,
					variants: ctx.variants,
				}
			}

			if (nodes.NodeType.CssDeclaration === node.type) {
				return {
					target: node,
					value: text.slice(node.expr.range[0], node.expr.range[1]),
					important: ctx.important || node.important,
					variants: ctx.variants,
				}
			}

			if (nodes.NodeType.ArbitraryClassname === node.type) {
				return {
					target: node,
					value: text.slice(node.range[0], node.range[1]),
					important: ctx.important || node.important,
					variants: ctx.variants,
				}
			}

			return undefined
		}

		if (node == undefined) {
			return undefined
		}

		if (nodes.NodeType.Program === node.type) {
			const expr = node.expressions.find(inRange)
			if (expr) return walk(expr)
			return undefined
		}

		if (inRange(node)) {
			return walk(node)
		}

		return undefined
	}

	return travel(parse({ text, start, end, breac: position }), {
		important: false,
		variants: [],
	})
}
