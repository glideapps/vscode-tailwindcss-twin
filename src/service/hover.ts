import { ExtractedToken, ExtractedTokenKind, TextDocument } from "@/extractors"
import { defaultLogger as console } from "@/logger"
import parseThemeValue from "@/parseThemeValue"
import { transformSourceMap } from "@/sourcemap"
import * as parser from "@/twin-parser"
import { cssDataManager, getEntryDescription } from "@/vscode-css-languageservice"
import vscode from "vscode"
import type { ServiceOptions } from "."
import { getDescription, getReferenceLinks } from "./referenceLink"
import type { TailwindLoader } from "./tailwind"

export default async function hover(
	result: ExtractedToken | undefined,
	document: TextDocument,
	position: unknown,
	state: TailwindLoader,
	options: ServiceOptions,
	tabSize: number,
): Promise<vscode.Hover | undefined> {
	if (!result) return undefined

	return doHover(result)

	function doHover(result: ExtractedToken) {
		try {
			const { token, kind } = result
			if (kind === ExtractedTokenKind.TwinTheme) {
				const range = new vscode.Range(document.positionAt(token.start), document.positionAt(token.end))
				return resolveThemeValue({ kind, range, token, state, options })
			} else if (kind === ExtractedTokenKind.TwinScreen) {
				const range = new vscode.Range(document.positionAt(token.start), document.positionAt(token.end))
				return resolveScreenValue({ kind, range, token, state, options })
			} else {
				const selection = parser.hover({
					text: token.value,
					position: document.offsetAt(position) - token.start,
					separator: state.separator,
				})
				if (!selection) return undefined

				const { start, end } = selection.target
				let value = selection.target.value

				const range = new vscode.Range(
					document.positionAt(token.start + start),
					document.positionAt(token.start + end),
				)

				if (selection.type === parser.HoverResultType.CssProperty) {
					const prop = selection.prop?.toKebab() ?? ""
					const value = selection.value?.value ?? ""
					const important = selection.important

					const header = new vscode.MarkdownString()
					if (options.references) {
						const entry = cssDataManager.getProperty(prop)
						if (entry) {
							const desc = getEntryDescription(entry, true)
							if (desc) {
								header.appendMarkdown(desc.value)
							}
						}
					}

					const code = state.tw.renderCssProperty({
						prop,
						value,
						important,
						rootFontSize: options.rootFontSize,
						tabSize,
					})
					const codes = new vscode.MarkdownString()
					if (code) codes.appendCodeblock(code, "scss")

					if (!header.value && !codes.value) return undefined

					return {
						range,
						contents: [header, codes],
					}
				}

				if (kind !== ExtractedTokenKind.Twin) return undefined

				if (selection.type === parser.HoverResultType.ArbitraryVariant) {
					const header = new vscode.MarkdownString("**arbitrary variant**")
					const codes = new vscode.MarkdownString()
					let code = selection.target.value.slice(1, -1)
					if (!code) {
						return {
							range,
							contents: [header, codes],
						}
					}
					code = state.tw.renderArbitraryVariant(code, tabSize)
					if (code) codes.appendCodeblock(code, "scss")
					return {
						range,
						contents: [header, codes],
					}
				}

				if (selection.type === parser.HoverResultType.Variant) {
					const header = new vscode.MarkdownString()
					if (options.references) {
						const desc =
							state.tw.screens.indexOf(value) === -1 ? getDescription(value) : getDescription("screens")
						if (typeof desc === "string") {
							header.appendMarkdown(desc ? desc + "\n" : "twin.marco" + "\n")
						}

						const links = getReferenceLinks(value)

						if (links.length > 0) {
							header.appendMarkdown("\n")
							header.appendMarkdown(links.map(ref => `[Reference](${ref.url}) `).join("\n"))
						}
					}

					const code = state.tw.renderVariant(value, tabSize)
					const codes = new vscode.MarkdownString()
					if (code) codes.appendCodeblock(code, "scss")

					if (!header.value && !codes.value) return undefined

					return {
						range,
						contents: [header, codes],
					}
				}

				const header = new vscode.MarkdownString()
				if (options.references) {
					const plugin = state.tw.getPlugin(value)
					let name = state.tw.trimPrefix(value)
					if (plugin) name = plugin.name
					if (name) {
						const desc = getDescription(name)
						if (typeof desc === "string") {
							header.appendMarkdown(desc ? desc + "\n" : "twin.marco" + "\n")
						}

						const links = getReferenceLinks(name)
						if (links.length > 0) {
							header.appendMarkdown("\n")
							header.appendMarkdown(links.map(ref => `[Reference](${ref.url}) `).join("\n"))
						}
					}
				}

				let code = state.tw.renderClassname({
					classname: value,
					important: selection.important,
					rootFontSize: options.rootFontSize,
					tabSize,
				})

				if (!code) {
					const i = value.lastIndexOf("/")
					const n = value.charCodeAt(i + 1)
					if (i !== -1 && (n === 91 || (n >= 48 && n <= 57))) {
						value = value.slice(0, i)
					}
					code = state.tw.renderClassname({
						classname: value,
						important: selection.important,
						rootFontSize: options.rootFontSize,
						tabSize,
					})
				}

				const codes = new vscode.MarkdownString()
				if (code) codes.appendCodeblock(code, "css")

				if (!header.value && !codes.value) return undefined

				return {
					range,
					contents: [header, codes],
				}
			}
		} catch (error) {
			const err = error as Error
			if (err.stack) err.stack = transformSourceMap(options.serverSourceMapUri.fsPath, err.stack)
			console.error(err)
			console.error("hover failed.")
		}

		return undefined
	}
}

function resolveThemeValue({
	range,
	token,
	state,
}: {
	kind: ExtractedTokenKind
	range: vscode.Range
	token: parser.Token
	state: TailwindLoader
	options: ServiceOptions
}): vscode.Hover | undefined {
	const result = parseThemeValue(token.value)
	if (result.errors.length > 0) {
		return undefined
	}

	const value = state.tw.getTheme(result.keys(), true)

	const markdown = new vscode.MarkdownString()

	if (typeof value === "string") {
		markdown.value = `\`\`\`txt\n${value}\n\`\`\``
	} else if (value instanceof Array) {
		markdown.value = `\`\`\`txt\n${value.join(", ")}\n\`\`\``
	} else if (value) {
		markdown.value = `\`\`\`js\n${value.toString?.() ?? typeof value}\n\`\`\``
	}

	return {
		range,
		contents: [markdown],
	}
}

function resolveScreenValue({
	range,
	token,
	state,
}: {
	kind: ExtractedTokenKind
	range: vscode.Range
	token: parser.Token
	state: TailwindLoader
	options: ServiceOptions
}): vscode.Hover | undefined {
	const value = state.tw.getTheme(["screens", token.value])
	if (value == undefined) {
		return
	}

	const markdown = new vscode.MarkdownString()

	if (typeof value === "string") {
		markdown.value = `\`\`\`css\n@media (min-width: ${value})\n\`\`\``
	} else if (value instanceof Array) {
		markdown.value = `\`\`\`txt\n${value.join(", ")}\n\`\`\``
	} else if (value) {
		markdown.value = `\`\`\`js\n${value.toString?.() ?? typeof value}\n\`\`\``
	}

	return {
		range,
		contents: [markdown],
	}
}
