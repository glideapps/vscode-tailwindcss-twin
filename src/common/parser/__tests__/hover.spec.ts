import { NodeType } from ".."
import { hover } from "../hover"

it("parser util hover", () => {
	const code = `bg-gray-500 padding[ 3rem ]!
	hover:(border-4 border-blue-500)
	[
		.group:hover,
		// .group:focus]
	]:!(text-red-700)// bg-primary
	/* text-error ()
	*/()
	custom/30 text-yellow-600/80 bg-gray-500/[.3]
	bg-[url(http://example)]
	text-[rgba(3 3 3 / .8)]/[.2] sm:hover: ( q-test`

	expect(
		hover({
			text: code,
			position: 32,
		}),
	).toEqual({
		target: {
			type: NodeType.SimpleVariant,
			range: [30, 36],
			id: {
				type: NodeType.Identifier,
				value: "hover",
				range: [30, 35],
			},
		},
		important: false,
		variants: [],
		value: "hover",
	})

	expect(
		hover({
			text: code,
			position: 117,
		}),
	).toMatchSnapshot()
})
