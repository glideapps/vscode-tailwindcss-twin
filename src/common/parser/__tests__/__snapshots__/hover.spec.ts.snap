// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser util hover 1`] = `
Object {
  "important": true,
  "target": Object {
    "important": false,
    "range": Array [
      106,
      118,
    ],
    "type": "ClassName",
    "value": "text-red-700",
  },
  "value": "text-red-700",
  "variants": Array [
    "
		.group:hover,
		// .group:focus]
	",
  ],
}
`;
